wallet/src/actors/app/methods.rs

```diff
@@ -445,31 +446,40 @@ impl App {
     }
 
     /// Handle new block notifications received from the node.
-    pub fn handle_block_notification(&self, value: types::Json) -> Result<()> {
-        log::trace!("received block notification");
+    pub fn handle_block_notification(&mut self, value: types::Json) -> Result<()> {
         let block = serde_json::from_value::<types::ChainBlock>(value).map_err(node_error)?;
 
-        for (_id, wallet) in self
+        // This iterator is collected early so as to free the immutable reference to `self`.
```

Comment:
This is because `self.state.get_sink` takes a mutable reference to self. No idea about what's the current architecture, but if every wallet has a sink, why not store it in the wallet instead of in self.state?

Reply:
It was inside `self.state.wallets` before, which was even crazier, lol.

Whoever wrote all that part of the wallet left no hints at all at why we got `Wallet`, `SessionWallet`, `State`, `Params` and a myriad more data structures that resemble each other without offering any additional convenience :shrug: